lib/parser/source/buffer.rb

```diff
@@ -253,14 +253,11 @@ def last_line
 
       def line_begins
         unless @line_begins
-          @line_begins, index = [ [ 0, 0 ] ], 1
-
-          @source.each_char do |char|
-            if char == NEW_LINE
-              @line_begins.unshift [ @line_begins.length, index ]
-            end
+          @line_begins, index = [ [ 0, 0 ] ], 0
 
+          while index = @source.index("\n".freeze, index)
```

Comment:
`"string literal".freeze` compiles to a special opcode in MRI, which uses a single, shared object. It doesn't actually call `Object#freeze` every time.